### DISCUSSION TITLE
gstreamer nvvidconv failing

### DISCUSSION THREAD

**Discussion by DaveBubenik** [OE4T/tegra-demo-distro#192] [Q&A] [answered]
I'm having trouble with a gstreamer command. My command works using the Linux distro release by NVIDIA, but I can't get it to run with this distro.

The following pipeline executes fine on this distro:

> gst-launch-1.0 v4l2src ! videoconvert ! omxh264enc bitrate=10000000 ! fakesink
Setting pipeline to PAUSED ...
Pipeline is live and does not need PREROLL ...
Setting pipeline to PLAYING ...
New clock: GstSystemClock
Framerate set to : 60 at NvxVideoEncoderSetParameterNvMMLiteOpen : Block : BlockType = 4
===== NVMEDIA: NVENC =====
NvMMLiteBlockCreate : Block : BlockType = 4
H264: Profile = 66, Level = 40
NVMEDIA_ENC: bBlitMode is set to TRUE
^Chandling interrupt.
Interrupt: Stopping pipeline ...
Execution ended after 0:00:01.812403837
Setting pipeline to NULL ...
Freeing pipeline ...

But when I try to use nvvidconv, I get errors:
>gst-launch-1.0 v4l2src ! nvvidconv ! omxh264enc bitrate=10000000 ! fakesink
Setting pipeline to PAUSED ...
Pipeline is live and does not need PREROLL ...
Setting pipeline to PLAYING ...
New clock: GstSystemClock
Framerate set to : 60 at NvxVideoEncoderSetParameterNvMMLiteOpen : Block : BlockType = 4
===== NVMEDIA: NVENC =====
NvMMLiteBlockCreate : Block : BlockType = 4
H264: Profile = 66, Level = 40
Caught SIGSEGV
exec gdb failed: No such file or directory
Spinning.  Please run 'gdb gst-launch-1.0 5221' to continue debugging, Ctrl-C to quit, or Ctrl-\ to dump core.

Could I be missing something on my image? Is there a certain package I need to include?

**Comment by kekiefer** [comment, marked as answer]
I suspect this has something to do with the fixes to buffer sizes to improve compatibility in newer versions of gstreamer not playing nice with the openmax encoder. Perhaps a workaround that may even be a better way of doing it is to use the newer `nvv4l2h264enc` plugin, which doesn't suffer the same issue.

**Reply by DaveBubenik** [reply]
Looks like that plugin worked! For anyone else reading this post, I'm using dunfell gstreamer 1.16.

**Comment by madisongh** [comment]
Which branch are you using, and which version of gstreamer?  I've just tried that pipeline on builds off master and dunfell, using gstreamer 1.14, and I'm not seeing the segfault.

**Reply by DaveBubenik** [reply]
I'm using dunfell gstreamer 1.16.  The nvv4l2h264enc plugin works, so I'm going with that.

**Reply by madisongh** [reply]
Yep, I just fired up 1.16 and ran into the same problem, and it's definitely due some buffer size mismatches.

**Reply by DaveBubenik** [reply]
I noticed meta-tegra has some bitbake files for gsteamer 1.18. Maybe I should use that instead? We need to use a later version for features that don't exist in 1.14.

**Reply by madisongh** [reply]
The 1.18 recipes are for branches where 1.18 is present in OE-Core.  If you're using dunfell, 1.16 is the right version.  In many cases, plugins written for older versions of gstreamer remain compatible with newer versions of gstreamer.

I've opened [an issue](https://github.com/OE4T/meta-tegra/issues/907) to track the problem with the OpenMAX plugin.

**Reply by DaveBubenik** [reply]
We don't have a strict requirement to use dunfell. Perhaps I should be using another branch? Any recommendations?

**Reply by madisongh** [reply]
dunfell is a good choice for maximum compatibility between L4T and OE-Core, plus it's the current long-term stable branch for OE-Core, so it will have support for a while yet.

It also depends on whether you'll be using NVIDIA SDKs. They're more likely to work on dunfell builds (with gstreamer 1.14), since the NV-provided binaries are built on the relatively ancient versions of glibc, gstreamer, etc. that are in Ubuntu 18.04.

If you're still early in development, you could use honister or even master, although there are some compatibility issues in master that we have yet to sort out by the time the next long-term stable OE-Core branch (kirkstone) is released in the next couple of months.  (And by "sort out," in some cases the answer will be "known issue, can't fix," unfortunately.)
